src/session/content/message_row/document.rs

```diff
@@ -0,0 +1,336 @@
+            Downloading(progress)
+        } else if remote.is_uploading_active {
+            let progress = remote.uploaded_size as f64 / size as f64;
+            Uploading(progress)
+        } else if local.is_downloading_completed {
+            Downloaded
+        } else if local.can_be_downloaded {
+            CanBeDownloaded
+        } else {
+            dbg!(file);
+            unimplemented!("unknown file status");
+        }
+    }
+}
+
+impl MessageDocument {
+    fn update_document(&self, message: &Message) {
+        if let MessageContent::MessageDocument(data) = message.content().0 {
+            let imp = self.imp();
+
+            let message_text = parse_formatted_text(data.caption);
+            imp.content_label.set_visible(!message_text.is_empty());
+            imp.content_label.set_label(message_text);
+
+            imp.file_name.set_label(&data.document.file_name);
+
+            imp.file_size.set_width_chars(16);
+
+            let session = message.chat().session();
+
+            self.update_status(data.document.document, session);
+        }
+    }
+
+    fn update_status(&self, file: File, session: Session) -> bool {
+        let status = FileStatus::from(&file);
+
+        let size = file.size.max(file.expected_size) as u64;
+
+        self.update_size_label(&status, size);
+        self.update_button(file, session, &status);
+
+        status == Downloaded
+    }
+
+    fn update_button(&self, file: File, session: Session, status: &FileStatus) {
+        let button = &self.imp().file_status;
+
+        let file_id = file.id;
+
+        let handler_id = match *status {
+            Downloading(progress) => {
+                // Cancel downloading
+                if progress > 0.1 {
+                    return;
+                }
+                button.set_icon_name("media-playback-stop-symbolic");
+                button.connect_clicked(clone!(@weak session => move |_| {
+                    session.cancel_download_file(file_id);
+                }))
+            }
+            Uploading(_progress) => {
+                // Cancel sending
+                unimplemented!();
+            }
+            CanBeDownloaded => {
+                // Download file
+                button.set_icon_name("document-save-symbolic");
+                button.connect_clicked(clone!( @weak self as obj, @weak session => move |_| {
+                        let (sender, receiver) = glib::MainContext::sync_channel::<File>(Default::default(), 5);
+                        receiver.attach(
+                            None,
+                            clone!(@weak obj, @weak session => @default-return glib::Continue(false), move |file| {
+                                let file_downloaded = obj.update_status(file, session);
+                                glib::Continue(!file_downloaded)
+                            }));
+
+                        session.download_file(file_id, sender);
+                    }))
```

Comment:
Besides indentation, this changes the clicked handler to `cancel` when the button has been clicked.  
```suggestion
                    let (sender, receiver) = glib::MainContext::sync_channel::<File>(Default::default(), 5);
                    receiver.attach(
                        None,
                        clone!(@weak obj, @weak session => @default-return glib::Continue(false), move |file| {
                            let file_downloaded = obj.update_status(file, session);
                            glib::Continue(!file_downloaded)
                        }));

                    session.download_file(file_id, sender);

                    button.set_icon_name("media-playback-stop-symbolic");

                    let handler_id = button.connect_clicked(clone!(@weak session => move |_| {
                        session.cancel_download_file(file_id);
                    }));
                    if let Some(handler_id) = obj.imp().status_handler_id.replace(Some(handler_id)) {
                        button.disconnect(handler_id);
                    }
```